File: src/test/java/no/fintlabs/ManualEventCreationServiceTest.java
package no.fintlabs;

import no.fintlabs.exceptions.LatestStatusEventNotOfTypeErrorException;
import no.fintlabs.exceptions.NoPreviousStatusEventsFoundException;
import no.fintlabs.flyt.kafka.instanceflow.headers.InstanceFlowHeaders;
import no.fintlabs.model.SourceApplicationAggregateInstanceId;
import no.fintlabs.model.event.Event;
import no.fintlabs.model.event.EventCategory;
import no.fintlabs.model.event.EventType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Clock;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;

class ManualEventCreationServiceTest {

    private final Clock clock = Clock.fixed(Instant.ofEpochMilli(100000), ZoneOffset.UTC);
    private UuidService uuidService;
    private EventService eventService;
    private ManualEventCreationService manualEventCreationService;

    @BeforeEach
    void setup() {
        uuidService = mock(UuidService.class);
        eventService = mock(EventService.class);
        manualEventCreationService = new ManualEventCreationService(
                clock,
                uuidService,
                "testApplicationId",
                eventService
        );
    }

    @Test
    public void givenNoPreviousEventForAggregateSourceApplicationId_whenSave_thenThrowException() {
        SourceApplicationAggregateInstanceId sourceApplicationAggregateInstanceId =
                new SourceApplicationAggregateInstanceId() {
                    @Override
                    public Long getSourceApplicationId() {
                        return 1L;
                    }

                    @Override
                    public String getSourceApplicationIntegrationId() {
                        return "testSourceApplicationIntegrationId";
                    }

                    @Override
                    public String getSourceApplicationInstanceId() {
                        return "testSourceApplicationInstanceId";
                    }
                };

        EventCategory eventCategory = EventCategory.INSTANCE_RECEIVED;

        when(eventService.findLatestStatusEventBySourceApplicationAggregateInstanceId(
                sourceApplicationAggregateInstanceId
        )).thenReturn(Optional.empty());

        assertThrows(
                NoPreviousStatusEventsFoundException.class,
                () -> manualEventCreationService.save(
                        sourceApplicationAggregateInstanceId,
                        eventCategory,
                        "testArchiveInstanceId"
                )
        );

        verify(eventService, times(1))
                .findLatestStatusEventBySourceApplicationAggregateInstanceId(
                        sourceApplicationAggregateInstanceId
                );

        verifyNoMoreInteractions(
                uuidService,
                eventService
        );
    }

    @Test
    public void givenPreviousEventForAggregateSourceApplicationIdNotOfTypeError_whenSave_thenThrowException() {
        SourceApplicationAggregateInstanceId sourceApplicationAggregateInstanceId =
                new SourceApplicationAggregateInstanceId() {
                    @Override
                    public Long getSourceApplicationId() {
                        return 1L;
                    }

                    @Override
                    public String getSourceApplicationIntegrationId() {
                        return "testSourceApplicationIntegrationId";
                    }

                    @Override
                    public String getSourceApplicationInstanceId() {
                        return "testSourceApplicationInstanceId";
                    }
                };

        EventCategory eventCategory = EventCategory.INSTANCE_RECEIVED;

        Event event = mock(Event.class);
        when(event.getType()).thenReturn(EventType.INFO);

        when(eventService.findLatestStatusEventBySourceApplicationAggregateInstanceId(
                sourceApplicationAggregateInstanceId
        )).thenReturn(Optional.of(event));

        assertThrows(
                LatestStatusEventNotOfTypeErrorException.class,
                () -> manualEventCreationService.save(
                        sourceApplicationAggregateInstanceId,
                        eventCategory,
                        "testArchiveInstanceId"
                )
        );

        verify(eventService, times(1))
                .findLatestStatusEventBySourceApplicationAggregateInstanceId(
                        sourceApplicationAggregateInstanceId
                );

        verifyNoMoreInteractions(
                uuidService,
                eventService
        );
    }

    @Test
    public void givenPreviousEventForAggregateSourceApplicationIdIsOfTypeInfo_whenSave_thenCallRepositorySaveAndReturnResult() {
        SourceApplicationAggregateInstanceId sourceApplicationAggregateInstanceId =
                new SourceApplicationAggregateInstanceId() {
                    @Override
                    public Long getSourceApplicationId() {
                        return 1L;
                    }

                    @Override
                    public String getSourceApplicationIntegrationId() {
                        return "testSourceApplicationIntegrationId";
                    }

                    @Override
                    public String getSourceApplicationInstanceId() {
                        return "testSourceApplicationInstanceId";
                    }
                };

        Event latestEvent = mock(Event.class);
        when(latestEvent.getType()).thenReturn(EventType.ERROR);
        InstanceFlowHeaders latestEventInstanceFlowHeaders = mock(InstanceFlowHeaders.class);
        when(latestEventInstanceFlowHeaders.getIntegrationId()).thenReturn(10L);
        when(latestEvent.getInstanceFlowHeaders()).thenReturn(latestEventInstanceFlowHeaders);

        when(eventService.findLatestStatusEventBySourceApplicationAggregateInstanceId(
                sourceApplicationAggregateInstanceId
        )).thenReturn(Optional.of(latestEvent));

        Event newEvent = mock(Event.class);

        UUID uuid = UUID.fromString("079c8300-8af7-4955-87bd-1efce87d52cd");
        when(uuidService.generateUuid()).thenReturn(uuid);

        when(eventService.save(argThat(e ->
                e.getInstanceFlowHeaders().equals(
                        InstanceFlowHeaders
                                .builder()
                                .sourceApplicationId(1L)
                                .sourceApplicationIntegrationId("testSourceApplicationIntegrationId")
                                .sourceApplicationInstanceId("testSourceApplicationInstanceId")
                                .integrationId(10L)
                                .correlationId(uuid)
                                .archiveInstanceId("testArchiveInstanceId")
                                .build()
                ) && e.getCategory().equals(EventCategory.INSTANCE_MANUALLY_PROCESSED)
                && e.getTimestamp().isEqual(OffsetDateTime.now(clock))
                && e.getType().equals(EventType.INFO)
                && e.getApplicationId().equals("testApplicationId")
        ))).thenReturn(newEvent);

        Event result = manualEventCreationService.save(
                sourceApplicationAggregateInstanceId,
                EventCategory.INSTANCE_MANUALLY_PROCESSED,
                "testArchiveInstanceId"
        );

        verify(eventService, times(1))
                .findLatestStatusEventBySourceApplicationAggregateInstanceId(
                        sourceApplicationAggregateInstanceId
                );

        verify(uuidService, times(1)).generateUuid();

        verify(eventService, times(1)).save(argThat(e ->
                e.getInstanceFlowHeaders().equals(
                        InstanceFlowHeaders
                                .builder()
                                .sourceApplicationId(1L)
                                .sourceApplicationIntegrationId("testSourceApplicationIntegrationId")
                                .sourceApplicationInstanceId("testSourceApplicationInstanceId")
                                .integrationId(10L)
                                .correlationId(uuid)
                                .archiveInstanceId("testArchiveInstanceId")
                                .build()
                ) && e.getCategory().equals(EventCategory.INSTANCE_MANUALLY_PROCESSED)
                && e.getTimestamp().isEqual(OffsetDateTime.now(clock))
                && e.getType().equals(EventType.INFO)
                && e.getApplicationId().equals("testApplicationId")
        ));

        verifyNoMoreInteractions(
                uuidService,
                eventService
        );

        assertThat(result).isEqualTo(newEvent);
    }

}

File: src/test/kotlin/no/novari/flyt/history/ManualEventCreationServiceTest.kt
package no.novari.flyt.history

import no.novari.flyt.history.exceptions.LatestStatusEventNotOfTypeErrorException
import no.novari.flyt.history.exceptions.NoPreviousStatusEventsFoundException
import no.novari.flyt.history.model.SourceApplicationAggregateInstanceId
import no.novari.flyt.history.model.event.Event
import no.novari.flyt.history.model.event.EventCategory
import no.novari.flyt.history.model.event.EventType
import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeaders
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.mockito.kotlin.argThat
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import java.time.Clock
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.UUID

class ManualEventCreationServiceTest {
    private val clock: Clock = Clock.fixed(Instant.ofEpochMilli(100000), ZoneOffset.UTC)
    private lateinit var uuidService: UuidService
    private lateinit var eventService: EventService
    private lateinit var manualEventCreationService: ManualEventCreationService

    @BeforeEach
    fun setup() {
        uuidService = mock()
        eventService = mock()
        manualEventCreationService =
            ManualEventCreationService(
                clock,
                uuidService,
                "testApplicationId",
                eventService,
            )
    }

    @Test
    fun `given no previous event for aggregate source application id when save then throw exception`() {
        val sourceApplicationAggregateInstanceId = testSourceApplicationAggregateInstanceId()
        whenever(
            eventService.findLatestStatusEventBySourceApplicationAggregateInstanceId(
                sourceApplicationAggregateInstanceId,
            ),
        ).thenReturn(null)

        assertThrows<NoPreviousStatusEventsFoundException> {
            manualEventCreationService.save(
                sourceApplicationAggregateInstanceId,
                EventCategory.INSTANCE_RECEIVED,
                "testArchiveInstanceId",
            )
        }

        verify(eventService, times(1))
            .findLatestStatusEventBySourceApplicationAggregateInstanceId(sourceApplicationAggregateInstanceId)
        verifyNoMoreInteractions(uuidService, eventService)
    }

    @Test
    fun `given previous event not of type error when save then throw exception`() {
        val sourceApplicationAggregateInstanceId = testSourceApplicationAggregateInstanceId()
        val event: Event = mock()
        whenever(event.type).thenReturn(EventType.INFO)
        whenever(
            eventService.findLatestStatusEventBySourceApplicationAggregateInstanceId(
                sourceApplicationAggregateInstanceId,
            ),
        ).thenReturn(event)

        assertThrows<LatestStatusEventNotOfTypeErrorException> {
            manualEventCreationService.save(
                sourceApplicationAggregateInstanceId,
                EventCategory.INSTANCE_RECEIVED,
                "testArchiveInstanceId",
            )
        }

        verify(eventService, times(1))
            .findLatestStatusEventBySourceApplicationAggregateInstanceId(sourceApplicationAggregateInstanceId)
        verifyNoMoreInteractions(uuidService, eventService)
    }

    @Test
    fun `given previous event is of type error when save then call event service save and return result`() {
        val sourceApplicationAggregateInstanceId = testSourceApplicationAggregateInstanceId()

        val latestEvent: Event = mock()
        whenever(latestEvent.type).thenReturn(EventType.ERROR)
        val latestEventInstanceFlowHeaders: InstanceFlowHeaders = mock()
        whenever(latestEventInstanceFlowHeaders.integrationId).thenReturn(10L)
        whenever(latestEvent.instanceFlowHeaders).thenReturn(latestEventInstanceFlowHeaders)
        whenever(
            eventService.findLatestStatusEventBySourceApplicationAggregateInstanceId(
                sourceApplicationAggregateInstanceId,
            ),
        ).thenReturn(latestEvent)

        val newEvent: Event = mock()
        val uuid = UUID.fromString("079c8300-8af7-4955-87bd-1efce87d52cd")
        whenever(uuidService.generateUuid()).thenReturn(uuid)
        whenever(
            eventService.save(
                argThat { eventMatches(this, uuid) },
            ),
        ).thenReturn(newEvent)

        val result =
            manualEventCreationService.save(
                sourceApplicationAggregateInstanceId,
                EventCategory.INSTANCE_MANUALLY_PROCESSED,
                "testArchiveInstanceId",
            )

        verify(eventService, times(1))
            .findLatestStatusEventBySourceApplicationAggregateInstanceId(sourceApplicationAggregateInstanceId)
        verify(uuidService, times(1)).generateUuid()
        verify(eventService, times(1)).save(argThat { eventMatches(this, uuid) })
        verifyNoMoreInteractions(uuidService, eventService)
        assertThat(result).isEqualTo(newEvent)
    }

    private fun testSourceApplicationAggregateInstanceId(): SourceApplicationAggregateInstanceId {
        return object : SourceApplicationAggregateInstanceId {
            override val sourceApplicationId = 1L

            override val sourceApplicationIntegrationId = "testSourceApplicationIntegrationId"

            override val sourceApplicationInstanceId = "testSourceApplicationInstanceId"
        }
    }

    private fun eventMatches(
        event: Event,
        uuid: UUID,
    ): Boolean {
        return event.instanceFlowHeaders ==
            InstanceFlowHeaders
                .builder()
                .sourceApplicationId(1L)
                .sourceApplicationIntegrationId("testSourceApplicationIntegrationId")
                .sourceApplicationInstanceId("testSourceApplicationInstanceId")
                .integrationId(10L)
                .correlationId(uuid)
                .archiveInstanceId("testArchiveInstanceId")
                .build() &&
            event.category == EventCategory.INSTANCE_MANUALLY_PROCESSED &&
            event.timestamp!!.isEqual(OffsetDateTime.now(clock)) &&
            event.type == EventType.INFO &&
            event.applicationId == "testApplicationId"
    }
}
